picoquic/pacing.c

```diff
@@ -0,0 +1,303 @@
+*/
+int picoquic_is_authorized_by_pacing(picoquic_pacing_t * pacing, uint64_t current_time, uint64_t * next_time,
+    unsigned int packet_train_mode, picoquic_quic_t * quic)
+{
+    int ret = 1;
+
+    picoquic_update_pacing_bucket(pacing, current_time);
+
+    if (pacing->bucket_nanosec < pacing->packet_time_nanosec) {
+        uint64_t next_pacing_time;
+        int64_t bucket_required;
+
+        if (packet_train_mode || pacing->bandwidth_pause) {
+            bucket_required = pacing->bucket_max;
+
+            if (bucket_required > 10 * pacing->packet_time_nanosec) {
+                bucket_required = 10 * pacing->packet_time_nanosec;
+            }
+
+            bucket_required -= pacing->bucket_nanosec;
+        }
+        else {
+            bucket_required = pacing->packet_time_nanosec - pacing->bucket_nanosec;
+        }
+
+        next_pacing_time = current_time + 1 + bucket_required / 1000;
+        if (next_pacing_time < *next_time) {
+            pacing->bandwidth_pause = 0;
+            *next_time = next_pacing_time;
+            if (quic != NULL) {
+                SET_LAST_WAKE(quic, PICOQUIC_SENDER);
+            }
+        }
+        ret = 0;
+    }
+
+    return ret;
+}
+
+/* Report pacing updates if required
+ */
+static void picoquic_report_pacing_update(picoquic_pacing_t* pacing, picoquic_path_t* path_x)
+{
+    picoquic_cnx_t* cnx = path_x->cnx;
+
+    if (cnx->is_pacing_update_requested && path_x == cnx->path[0] &&
+        cnx->callback_fn != NULL) {
+        if ((pacing->rate > cnx->pacing_rate_signalled &&
+            (pacing->rate - cnx->pacing_rate_signalled >= cnx->pacing_increase_threshold)) ||
+            (pacing->rate < cnx->pacing_rate_signalled &&
+                (cnx->pacing_rate_signalled - pacing->rate > cnx->pacing_decrease_threshold))){
+            (void)cnx->callback_fn(cnx, pacing->rate, NULL, 0, picoquic_callback_pacing_changed, cnx->callback_ctx, NULL);
+            cnx->pacing_rate_signalled = pacing->rate;
+        }
+    }
+    if (cnx->is_path_quality_update_requested &&
+        cnx->callback_fn != NULL) {
+        /* TODO: add a function "export path quality" */
+        /* TODO: remember previous signalled value for change tests */
+        if (path_x->smoothed_rtt < path_x->rtt_threshold_low ||
+            path_x->smoothed_rtt > path_x->rtt_threshold_high ||
+            pacing->rate < path_x->pacing_rate_threshold_low ||
+            pacing->rate > path_x->pacing_rate_threshold_high) {
+            (void)cnx->callback_fn(cnx, path_x->unique_path_id, NULL, 0, picoquic_callback_path_quality_changed, cnx->callback_ctx, path_x->app_path_ctx);
+            picoquic_refresh_path_quality_thresholds(path_x);
+        }
+    }
+}
+
+/* Reset the pacing data after recomputing the pacing rate
+*/
+void picoquic_update_pacing_parameters(picoquic_pacing_t * pacing, double pacing_rate, uint64_t quantum, size_t send_mtu, uint64_t smoothed_rtt,
+    picoquic_path_t * signalled_path)
+{
+#if 0
```

Comment:
not sure if we should need this block anymore ?

Reply:
Yes, I will clean that.

Reply:
Cleaned.